### ISSUE TITLE
怎么使用grpo代码版本训练？

### ISSUE THREAD

**Issue by wangat** [modelscope/ms-swift#7049]
求助，目前测试命令行版本可以训练grpo，但是想确定张量变化情况和代码细节。尝试按照如下代码训练：

import os
os.environ['CUDA_VISIBLE_DEVICES'] = '4'
#os.environ['MAX_PIXELS'] = '1073296' #图片尺寸都相同，1024x1024

from swift.llm import (
    get_model_tokenizer, load_dataset, get_template, EncodePreprocessor,
    get_multimodal_target_regex, LazyLLMDataset
)
from swift.utils import get_logger, get_model_parameter_info, plot_images, seed_everything
from swift.tuners import Swift, LoraConfig
from swift.trainers.rlhf_arguments import DPOConfig, GRPOConfig, RewardConfig 
from swift.trainers import Seq2SeqTrainer, Seq2SeqTrainingArguments
from functools import partial
import torch
from swift.trainers import DPOTrainer, GRPOTrainer
from randomreword import RandomReward
from reward_test2 import grpoReward
from swift.trainers.rlhf_trainer.grpo_trainer import VLLMClient

logger = get_logger()
seed_everything(42)

model_id_or_path = '/map_data/wyd/model/checkpoint-26000'
system = None  
output_dir = 'output'
deepspeed_path = '/map_data/wyd/result/test_map_d1/scripts_tools/test_2/ds_zero2_config_2.json'

train_dataset = '/map_data/wyd/data/msg_official_road_info/all_json/for_grpo_rl_test/rl_4.json'
data_seed = 42
max_length = 12288
split_dataset_ratio = 0.1  # Split validation set
num_proc = 4  # The number of processes for data loading.

vllm_client = VLLMClient(
    base_urls=["http://172.22.65.172:8011"], # 确保端口与启动server时一致
    #hosts=['127.0.0.1'],
    #server_ports=[8011],
    connection_timeout=240,
)

training_args = GRPOConfig(
    output_dir=output_dir,
    learning_rate=1e-6,
    per_device_train_batch_size=2,
    per_device_eval_batch_size=4,
    gradient_checkpointing=True,
    weight_decay=0.1,
    lr_scheduler_type='cosine',
    warmup_ratio=0.05, 
    #report_to=['tensorboard'],
    report_to=['swanlab'],
    logging_first_step=True,
    save_strategy='steps',
    save_steps=100,
    eval_strategy='steps',
    eval_steps=100,
    gradient_accumulation_steps=8,
    # To observe the training results more quickly, this is set to 1 here. 
    # Under normal circumstances, a larger number should be used.
    num_train_epochs=1,
    metric_for_best_model='loss',
    save_total_limit=2,
    logging_steps=50,
    dataloader_num_workers=1,
    data_seed=data_seed,
    remove_unused_columns=False,
    num_generations = 2, 
    beta = 0.001, #不修改的是默认值！
    loss_type="grpo",
    ref_model_sync_steps=512,
    ref_model_mixup_alpha=0.6,
    max_prompt_length=4096, #需要测试
    max_completion_length=8192,
    temperature=1.0,
    top_p=0.9,
    top_k=50,
    repetition_penalty=1.0,
    use_vllm=True,
    vllm_mode='server',
    vllm_max_model_len=12288,
    max_model_len = 12288,
)

output_dir = os.path.abspath(os.path.expanduser(output_dir))
logger.info(f'output_dir: {output_dir}')

model_kwargs = {
    'attn_implementation': 'flash_attention_2',
}

model, processor = get_model_tokenizer(model_id_or_path, model_type='qwen2_5_vl', model_kwargs=model_kwargs)
logger.info(f'model_info: {model.model_info}')
template = get_template(model.model_meta.template, processor, default_system=system, max_length=max_length)
template.set_mode('train')
model.config.use_cache = False
template.tokenizer.padding_side = 'left'
template.tokenizer.pad_token = template.tokenizer.eos_token
if template.use_model:
    template.model = model

logger.info(f'model: {model}')
model_parameter_info = get_model_parameter_info(model)
logger.info(f'model_parameter_info: {model_parameter_info}')

train_dataset, val_dataset = load_dataset(train_dataset, split_dataset_ratio=split_dataset_ratio, num_proc=num_proc,
                                          seed=data_seed)

logger.info(f'train_dataset: {train_dataset}')
logger.info(f'val_dataset: {val_dataset}')
logger.info(f'train_dataset[0]: {train_dataset[0]}')

train_dataset = LazyLLMDataset(train_dataset, template.encode, random_state=data_seed)
val_dataset = LazyLLMDataset(val_dataset, template.encode, random_state=data_seed)

for name, param in model.named_parameters():
    if name.startswith('visual.') or 'vision' in name.lower():
        param.requires_grad = False
        logger.info(f"Froze visual parameter: {name}")

model.enable_input_require_grads()  # Compatible with gradient checkpointing
trainer = GRPOTrainer(
    model=model,
    args=training_args,
    data_collator=template.data_collator,
    train_dataset=train_dataset,
    eval_dataset=val_dataset,
    template=template,
    reward_funcs = [grpoReward],#实现reward函数，测试命令行正常训练，但指标上看有地方设置不合理；
    vllm_client=vllm_client,
)
    
trainer.train()

last_model_checkpoint = trainer.state.last_model_checkpoint
logger.info(f'last_model_checkpoint: {last_model_checkpoint}')

训练中报[WARNING:swift] Encoding failed for one sample; resampling a new input. 'messages'，多次尝试之后报错Failed to obtain a valid sample after multiple attempts. Consider increasing `max_length` or adjusting the `truncation_strategy` to avoid excessive truncation.定位是在swift/trainers/rlhf_trainer/grpo_trainer/GRPOTrainer中的
if self.template.truncation_strategy == 'raise':
    inputs = self.resample_encode_failed_inputs(inputs)
最后在def from_dict(cls, inputs: Dict[str, Any]) -> 'TemplateInputs':中判断了has_rejected_messages = inputs.get('rejected_messages') is not None等后续代码报错；(数据是按照'solution'为相关gt、没有传入reject这样制备的，做grounding任务);

尝试修改template = get_template(model.model_meta.template, processor, default_system=system, max_length=max_length)中增加truncation_strategy='left'（这个是看https://swift.readthedocs.io/zh-cn/v3.8/Instruction/%E5%91%BD%E4%BB%A4%E8%A1%8C%E5%8F%82%E6%95%B0.html中参数）
会报错：File "/usr/local/lib/python3.11/site-packages/swift/trainers/rlhf_trainer/grpo_trainer.py", line 2228, in <listcomp>all_messages = gather_object([inp['messages'] for inp in inputs])；感觉grpo训练中数据应该被特殊处理了；

之后查看命令行版本代码，数据处理是在swift/llm/train/sft/SwiftSft的run函数下面，使用的是的load_dataset读取（from ..dataset import load_dataset），之后在_encode_dataset中是is_grpo = hasattr(args, 'rlhf_type') and args.rlhf_type == 'grpo'做了判断，最后直接返回了datasets；但是后续代码较多，短时间不太确定细节，想问一下有没有非命令行相关文档。感谢。

**Comment by hjh0119** [comment]
maybe you can refer to this issue to debug https://github.com/modelscope/ms-swift/issues/6316

**Comment by wangat** [comment]
> maybe you can refer to this issue to debug [#6316](https://github.com/modelscope/ms-swift/issues/6316)

Thank you very much for your reply, it has been very helpful. I will study the relevant code.
